nixos/gitea: drop mailerUseSendmail option and use PROTOCOL as an indication instead

SuperSandro2000 opened the pull request:
see #384535 , similar to #384580

The option was just introduced a few days back, so we don't need a removal notice.

This way is superior over the other PR as we do not have an option users need to turn which can be fully avoided. See commits for more details.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

dotlambda commented on nixos/modules/services/misc/gitea.nix:
```diff
@@ -21,6 +21,9 @@ let
     ${optionalString (cfg.extraConfig != null) cfg.extraConfig}
   '';
+
+  inherit (cfg.settings) mailer;
+  useSendmail = (mailer ? ENABLED) && mailer.ENABLED && (mailer ? SENDMAIL_PATH);
```
```suggestion
  inherit (cfg.settings) mailer;
  useSendmail = mailer.ENABLED or false && mailer.PROTOCOL or null == "sendmail";
```

SuperSandro2000 committed:
nixos/gitea: drop mailerUseSendmail option and use PROTOCOL as an indication instead

This option does not configure sendmail itself because it is impossible
as sendmail is an alias for many things and could mean msmtp or postfix
or exim or something else.

Instead we rely on the PROTOCOL setting as initially proposed #384582
and based on that open up the sandboxing settings because if the user
configures sendmail, they want it to work and not have to configure yet
another things.

Also makes postfix specific things conditional on postfix being enabled
as msmtp does not need them.

Also we can set SENDMAIL_PATH unconditionally as every wrapper I am
aware of uses that path.

SuperSandro2000 committed:
nixos/gitea: align maintainers with package

alyssais commented on nixos/modules/services/misc/gitea.nix:
```diff
@@ -906,7 +899,7 @@ in
           "AF_INET"
           "AF_INET6"
-        ] ++ optional cfg.mailerUseSendmail "AF_NETLINK";
+        ] ++ optional (useSendmail && config.services.postfix.enable) "AF_NETLINK";
```
Then it might not need netlink, and if it turns out to we can enable it.

alyssais reviewed (approved):
Approach looks fine to me. Don't use the software so can't tell if everything is set right.

SuperSandro2000 commented on nixos/modules/services/misc/gitea.nix:
```diff
@@ -664,13 +658,12 @@ in
 
         mailer = mkMerge [
+          {
+            SENDMAIL_PATH = lib.mkDefault "/run/wrappers/bin/sendmail";
```
We can put it behind a lib.mkIf but I am using it msmtp and nullmailer actually and they both use external mailserver.

Izorkin commented on nixos/modules/services/misc/gitea.nix:
```diff
@@ -884,18 +877,18 @@ in
         PrivateTmp = true;
-        PrivateDevices = true;
-        PrivateUsers = optional (!cfg.mailerUseSendmail) true;
+        PrivateDevices = !useSendmail;
```
This variant?
```
  PrivateDevices = ++ optional (config.services.nullmailer.enable && config.services.nullmailer.setSendmail);
```

SuperSandro2000 replied:
Indeed we don't need it.

SuperSandro2000 commented:
We are using a freeform submodule, so I don't think anything needs to be changed there or what did you have in mind?

Izorkin commented:
> We are using a freeform submodule, so I don't think anything needs to be changed there or what did you have in mind?

No, it's fine. I was wrong. I thought using the example config might conflict with the added options in this PR.

It seems that `lib.types.enum` in `mailer.PROTOCOL` does not work. This variant does not cause error:
```
      "mailer" = {
        ENABLED = true;
        PROTOCOL = "sendmail-test";
      };
```

Izorkin commented:
> It seems that `lib.types.enum` in `mailer.PROTOCOL` does not work. This variant does not cause error:

No, everything is fine. Forgot to apply PR on the test server.

SuperSandro2000 commented:
If something is for someone not working, please post your setup and try to figure out why sendmail fails with strace and grepping for something like sendmail.

I tested this usually like:

```
strace -p $PID_OF_GITEA | rg sendmail -C 50
```

qbit commented:
This breaks my config: https://github.com/qbit/xin/blob/main/hosts/box/default.nix#L675-L695

```
error: A definition for option `services.gitea.settings.mailer.SENDMAIL_PATH'\'' is not of type `absolute path'\''. Definition values:
- In `/nix/store/7rbnmr95zp5b7nha1wydrg4rkkn3a5fd-source/nixos/modules/services/misc/gitea.nix'\'': "sendmail"
```

Seems that the absolute path check needs to be removed?
